File: gpt-4-turbo/stage2/kernels_construct_vector_length_clause.c
#include <stdio.h>
#include <stdlib.h>
#include <openacc.h>

#define SEED 12345
#define NUM_TEST_CALLS 10

int test1(){
    int err = 0;
    srand(SEED);
    const int vector_length = 64; // Example vector length
    const int num_elements = 1024; // Total number of elements to process
    float *a = (float*)malloc(sizeof(float) * num_elements);
    float *b = (float*)malloc(sizeof(float) * num_elements);

    // Initialize arrays
    for (int i = 0; i < num_elements; ++i) {
        a[i] = rand() / (float)RAND_MAX;
        b[i] = 0.0f;
    }

#pragma acc kernels vector_length(vector_length)
    for (int i = 0; i < num_elements; ++i) {
        b[i] = a[i] * a[i];
    }

    // Check results
    for (int i = 0; i < num_elements; ++i) {
        if (fabs(b[i] - (a[i] * a[i])) > 1e-5) {
            err++;
        }
    }

    free(a);
    free(b);

    return err;
}

int main(){
    int failcode = 0;
    int failed = 0;
    for (int x = 0; x < NUM_TEST_CALLS; ++x){
        failed += test1();
    }
    if (failed != 0){
        failcode += (1 << 0);
    }
    printf("Test %s\n", failcode == 0 ? "PASSED" : "FAILED");
    return failcode;
}

File: gpt-4-turbo/stage2/kernels_construct_vector_length_clause.f90
! Test for OpenACC kernels construct vector_length clause.
!T1:kernels construct vector_length clause,V:2.7-2.3
      LOGICAL FUNCTION test1()
        USE OPENACC
        IMPLICIT NONE
        INCLUDE "acc_testsuite.Fh"

        INTEGER, PARAMETER :: N = 1024
        REAL, DIMENSION(N) :: A, B, C
        INTEGER :: i, errors
        REAL, DIMENSION(N) :: expected_C
        errors = 0

        ! Initialize arrays
        DO i = 1, N
          A(i) = REAL(i)
          B(i) = REAL(N - i + 1)
          expected_C(i) = A(i) + B(i)
        END DO

        ! Apply kernels construct with vector_length clause
        !$acc kernels vector_length(128)
        DO i = 1, N
          C(i) = A(i) + B(i)
        END DO
        !$acc end kernels

        ! Verify results
        DO i = 1, N
          IF (C(i) /= expected_C(i)) THEN
            errors = errors + 1
          END IF
        END DO

        IF (errors == 0) THEN
          test1 = .FALSE.
        ELSE
          test1 = .TRUE.
        END IF
      END FUNCTION test1

      PROGRAM main
        USE OPENACC
        IMPLICIT NONE
        INTEGER :: failcode, testrun
        LOGICAL :: failed
        INCLUDE "acc_testsuite.Fh"
        
        EXTERNAL :: test1
        failed = .FALSE.
        failcode = 0

        DO testrun = 1, NUM_TEST_CALLS
          failed = failed .or. test1()
        END DO
        IF (failed) THEN
          failcode = failcode + 2 ** 0
          failed = .FALSE.
        END IF
        CALL EXIT (failcode)
      END PROGRAM